src/clp_ffi_js/ir/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,233 @@
+
+#include <clp/ErrorCode.hpp>
+#include <clp/ffi/encoding_methods.hpp>
+#include <clp/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/ir/LogEvent.hpp>
+#include <clp/ir/LogEventDeserializer.hpp>
+#include <clp/ir/types.hpp>
+#include <clp/streaming_compression/zstd/Decompressor.hpp>
+#include <clp/TraceableException.hpp>
+#include <emscripten/em_asm.h>
+#include <emscripten/val.h>
+#include <spdlog/spdlog.h>
+
+#include <clp_ffi_js/ClpJsException.hpp>
+#include <clp_ffi_js/constants.hpp>
+
+using namespace std::literals::string_literals;
+
+namespace clp_ffi_js::ir {
+auto ClpIrV1Decoder::create(emscripten::val const& data_array) -> ClpIrV1Decoder {
+    auto const length{data_array["length"].as<size_t>()};
+    SPDLOG_INFO("ClpIrV1Decoder::ClpIrV1Decoder() got buffer of length={}", length);
+
+    auto data_buffer{std::make_unique<char const[]>(length)};
+    emscripten::val::module_property("HEAPU8")
+            .call<void>("set", data_array, reinterpret_cast<uintptr_t>(data_buffer.get()));
+
+    auto const zstd_decompressor{std::make_shared<clp::streaming_compression::zstd::Decompressor>()
+    };
+    zstd_decompressor->open(data_buffer.get(), length);
+
+    bool is_four_bytes_encoding{true};
+    if (auto const err{
+                clp::ffi::ir_stream::get_encoding_type(*zstd_decompressor, is_four_bytes_encoding)
+        };
+        clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success != err)
+    {
+        SPDLOG_CRITICAL("Failed to decode encoding type, err={}", err);
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decode encoding type."
+        );
+    }
+
+    if (false == is_four_bytes_encoding) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Is not four byte encoding."
+        );
+    }
+    auto result{clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>::create(
+            *zstd_decompressor
+    )};
+    if (result.has_error()) {
+        auto const error_code{result.error()};
+        SPDLOG_CRITICAL(
+                "Failed to decompress: {}:{}",
+                error_code.category().name(),
+                error_code.message()
+        );
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decompress"
+        );
+    }
+
+    return ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()));
+}
+
+auto ClpIrV1Decoder::get_estimated_num_events() const -> size_t {
+    return m_log_events.size();
+}
+
+auto ClpIrV1Decoder::build_idx(size_t begin_idx, size_t end_idx) -> emscripten::val {
+    constexpr size_t cFullRangeEndIdx{0};
+    if (0 != begin_idx && cFullRangeEndIdx != end_idx) {
```

Comment:
* This should be `||`, right?
* If we're going to treat `0` as a special value indicating the end of the range, then we should document it in the docstring. That said, I think the more common approach in C++ is the max value of the type? E.g., `string::npos`.

Reply:
* Right
* It's documented in the docstring of `get_estimated_num_events`.
  > @return 0 (cFullRangeEndIdx) to indicate there is no event has been deserialized yet by `build_idx()`.

  The max value approach also make sense. Do you think it would be better if we return a max value (e.g. -1) in `get_estimated_num_events`?

Reply:
If we go with renaming `get_estimated_num_events` to `get_num_events_buffered`, then I think returning 0 when there are no events would be fine.